System/Random.hs

```diff
@@ -257,6 +193,95 @@ mutableByteArrayContentsCompat :: MutableByteArray s -> Ptr Word8
 -- buildWord64 w0 w1 = ((fromIntegral w1) `shiftL` 32) .|. (fromIntegral w0)
 -- :}
 
+-- $randomgen
+--
+-- [/Example for RNG Implementors:/]
+--
+-- Suppose you want to use a [permuted congruential
+-- generator](https://en.wikipedia.org/wiki/Permuted_congruential_generator)
+-- as the source of entropy. You can make it an instance of
+-- `RandomGen`:
+--
+-- >>> data PCGen = PCGen !Word64 !Word64
+--
+-- >>> :{
+-- let stepGen :: PCGen -> (Word32, PCGen)
+--     stepGen (PCGen state inc) = let
+--       newState = state * 6364136223846793005 + (inc .|. 1)
+--       xorShifted = fromIntegral (((state `shiftR` 18) `xor` state) `shiftR` 27) :: Word32
+--       rot = fromIntegral (state `shiftR` 59) :: Word32
+--       out = (xorShifted `shiftR` (fromIntegral rot)) .|. (xorShifted `shiftL` fromIntegral ((-rot) .&. 31))
+--       in (out, PCGen newState inc)
+-- :}
+--
+-- >>> fst $ stepGen $ snd $ stepGen (PCGen 17 29)
+-- 3288430965
+--
+-- >>> :{
+-- instance RandomGen PCGen where
+--   next g = (fromIntegral y, h)
+--     where
+--       (y, h) = stepGen g
+--   split _ = error "This PRNG is not splittable"
+-- :}
+--
+-- Importantly, this implementation will not be as efficient as it
+-- could be because the random values are converted to 'Integer' and
+-- then to desired type.
+--
+-- Instead we should define (where e.g. @unBuildWord32 :: Word32 ->
+-- (Word16, Word16)@ is a function to pull apart a 'Word32' into a
+-- pair of 'Word16'):
+--
+-- >>> data PCGen' = PCGen' !Word64 !Word64
+--
+-- >>> :{
+-- instance RandomGen PCGen' where
+--   genWord8  (PCGen' s i) = (z, PCGen' s' i')
+--     where
+--       (x, PCGen s' i') = stepGen (PCGen s i)
+--       y = fst $ unBuildWord32 x
+--       z = fst $ unBuildWord16 y
+--   genWord16 (PCGen' s i) = (y, PCGen' s' i')
+--     where
+--       (x, PCGen s' i') = stepGen (PCGen s i)
+--       y = fst $ unBuildWord32 x
```

Comment:
Are these `genWord8` and `genWord16` implementations more performant than the default ones? If no, let's remove them from the example. If yes, perhaps we can improve the default implementations?